state-chain/pallets/cf-environment/src/lib.rs

```diff
@@ -458,17 +457,70 @@ pub mod pallet {
 		) -> DispatchResultWithPostInfo {
 			T::EnsureWitnessed::ensure_origin(origin)?;
 
-			if let Some(_nonce) = SolanaUnavailableNonceAccounts::<T>::take(nonce_account) {
+			if SolanaUnavailableNonceAccounts::<T>::take(nonce_account).is_some() {
 				SolanaAvailableNonceAccounts::<T>::append((nonce_account, durable_nonce));
 				Self::deposit_event(Event::<T>::DurableNonceSetForAccount {
 					nonce_account,
 					durable_nonce,
 				});
 				Ok(().into())
 			} else {
-				Err(Error::<T>::NonceAccountNotBeingUsedOrDoesntExist.into())
+				Err(Error::<T>::NonceAccountNotBeingUsedOrDoesNotExist.into())
 			}
 		}
+
+		/// Allows Governance to recover a used Nonce.
+		/// If a Hash is supplied as well, update the associated Durable Hash as well.
+		/// Requires Governance Origin.
+		///
+		/// ## Events
+		///
+		/// - [DurableNonceSetForAccount](Event::DurableNonceSetForAccount)
+		///
+		/// ## Errors
+		///
+		/// - [BadOrigin](frame_support::error::BadOrigin)
+		/// - [NonceAccountNotBeingUsed](Error::NonceAccountNotBeingUsedOrDoesNotExist)
+		/// - [NonceHashNotSuppliedWhereRequired](Error::NonceAccountNotBeingUsedOrDoesNotExist)
+		#[pallet::call_index(8)]
+		#[pallet::weight(Weight::zero())]
+		pub fn force_recover_sol_nonce(
+			origin: OriginFor<T>,
+			nonce_account: SolAddress,
+			durable_nonce: Option<SolHash>,
+		) -> DispatchResultWithPostInfo {
+			T::EnsureGovernance::ensure_origin(origin)?;
+
+			let new_hash =
+				// If Nonce account is currently Unavailable - reset it as Available again.
+				if let Some(current_hash) = SolanaUnavailableNonceAccounts::<T>::take(nonce_account) {
+					let new_hash = durable_nonce.unwrap_or(current_hash);
+					SolanaAvailableNonceAccounts::<T>::append((nonce_account, new_hash));
+					Ok(new_hash)
+				} else if let Some(new_hash) = durable_nonce {
+					// If the Nonce account is currently Available, update its Hash (therefore the Hash must be passed in)
+					SolanaAvailableNonceAccounts::<T>::try_mutate(|durable_nonces|{
+						durable_nonces
+							.iter()
+							.position(|(account, _)|*account == nonce_account)
+							.map(|idx| {
+								durable_nonces[idx] = (nonce_account, new_hash);
+								new_hash
+							})
```

Comment:
`iter_mut` is generally a better idea than using indices do update a map mid-iteration (even though I think it's ok here).

```rust
durable_nonces
	.iter_mut()
	.find(|(account, _)|*account == nonce_account)
	.map(|(_account, hash)| {
		*hash = new_hash;
		new_hash
	})
	.ok_or::<DispatchError>(Error::<T>::NonceAccountNotBeingUsedOrDoesNotExist.into())
```